### PR TITLE
Fix dependency handling on move or rename in the filesystem dock

### LINKED ISSUE #59116: Godot editor sometimes crashes when resources are moved into another folder
### Godot version

3.4.2

### System information

Windows 11, RTX 3080, GLES3

### Issue description

Description:
When I created some scenes, scripts etc and I want to rearrange them in my project structure. For instance move a scene from the root folder into my scene folder the editor crashes quite often. Not always but quite often.

Does the editor produces some error log? Maybe I can attach it then. I am quite new to Godot.

Operating System:
Windows 11 21H2 Build 22000.438

Godot version:
 3.4.2.stable

Thanks Bernd

### Steps to reproduce

Move a resource in the file system of the godot editor from the root folder to a subfolder. It crashes estimated every third time.

### Minimal reproduction project

_No response_

### LINKED ISSUE #77289: Packed scene files sometimes get "corrupted" when moving files within editor
### Godot version

4.0.3

### System information

Windows 11

### Issue description

I reorganised a project by dragging some files (various types including .tscn, .tres, .gd, images) into subdirectories within the project's filetree. This was all done within the editor.

After doing this I found two scene files would not open, giving the alarming popup message "scene file is invalid/corrupt".

After panicked googling and deleting the .godot directory (which did not fix the problem), I found that the problem was that scene A (the scene named in the message) contained an instance of scene B, and in scene B's .tscn file, the root node had been saved with a name but no type, i.e. instead of the line in the .tscn file saying:

`[node name="PlayerScene" type="ActorScene" groups=["players"]]`

It said:

`[node name="PlayerScene"]`
 
So Godot seems to have "lost" the root node's type when moving the scene file. Manually typing in the missing info enabled the file to load. The other "corrupt" scene depended on this scene, so it now loaded as well.

This seems like a really serious bug, as it means Godot's management of the project filetree is untrustworthy. It's worsened by the fact that the message produced by the editor is so opaque and unhelpful. It doesn't even identify the file causing the problem (scene B).

 

### Steps to reproduce

Not sure but I think it would involve having a scene that contains child nodes which are instances of other scenes, then moving the .tscn files to other directories.

### Minimal reproduction project

N/A

### LINKED ISSUE #79850: Crashing and breaking a scene when moved with parent scene
### Godot version

4.1.1

### System information

Godot v4.1.1.stable - Ubuntu 22.04.2 LTS 22.04 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 3GB (nvidia; 525.125.06) - AMD Ryzen 9 3900X 12-Core Processor (24 Threads)

### Issue description

If I move inherited scene, with the scene that I had inherited from + if those two scenes are open in the editor, it crashes, and the inherited scene is breaking. 

### Steps to reproduce

image
The project is already set up, the only thing that you need to do, is to open two scenes and then select them both and try to drug them to "directory" folder, and it should crash. 

### Minimal reproduction project

[Bug report.zip](https://github.com/godotengine/godot/files/12144957/Bug.report.zip)

### PR DESCRIPTION
**Pull request by Jordyfel**
Fixes #79850.
...and probably more.

Moving a scene's dependency in the filesystem while the scene is open should no longer corrupt the scene.
Moving a scene together with it's dependency while it's open should no longer crash the engine.
Also fixes most of the error spam that was happening when moving things around.

Open scenes which depend on moved files are now saved before the move, which prevents losing unsaved changes like before, but in a less error-prone way.
The `ResourceUID` system is properly informed of resources changing paths now, this was the source of most of the problems.
A lot of expensive function calls were removed, moving things around in the filesystem should be faster.
There are some errors remaining related to scripts being looked for in the old location, but nothing seems to actually break now.

Please test this by moving stuff around in the filesystem with a copy of your non-trivial project.

*Bugsquad edit:* Fixes #77289 Fixes #59116

### CODE REVIEW

File: editor/filesystem_dock.cpp [hunk 1]
```diff
@@ -1533,7 +1533,12 @@ void FileSystemDock::_try_duplicate_item(const FileOrFolder &p_item, const Strin
 	}
 }
 
-void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, String> &p_renames) const {
+void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, String> &p_renames, const HashMap<String, ResourceUID::ID> p_uids) const {
+	// Update the paths in ResourceUID, so that UIDs remain valid.
+	for (KeyValue pair : p_uids) {
```
[the remark this answers is not in the document]
Comment (Jordyfel) [reply editor/filesystem_dock.cpp:1538]: Thanks

File: editor/filesystem_dock.cpp [hunk 2]
```diff
@@ -1967,15 +1943,28 @@ void FileSystemDock::_move_operation_confirm(const String &p_to_path, bool p_cop
...
 			current_path_line_edit->set_text(current_path);
 		}
 	}
 }
 
+void FileSystemDock::_before_move(Vector<String> &r_old_paths, HashMap<String, ResourceUID::ID> &r_uids, Vector<String> &r_remaps) const {
+	for (int i = 0; i < to_move.size(); i++) {
+		r_old_paths.push_back(to_move[i].path);
+		ResourceUID::ID uid = ResourceLoader::get_resource_uid(to_move[i].path);
+		if (uid > 0) {
```
Comment (AThousandShips) [inline editor/filesystem_dock.cpp:1956]:
```suggestion
		if (uid != ResourceUID::INVALID_ID) {
```

File: editor/filesystem_dock.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jordyfel) [reply editor/filesystem_dock.cpp]: Making it a reference makes it complain with some error that I do not understand

[the remark this answers is not in the document]
Comment (AThousandShips) [reply editor/filesystem_dock.cpp]: Oh it should be KeyValue<String, ResourceUID::ID>

[the remark this answers is not in the document]
Comment (Jordyfel) [reply editor/filesystem_dock.cpp]: c++ is hard

**Comment by KoBeWi** [comment]
Seems like this fixes all crashes and errors when moving scenes (I didn't do super extensive testing though).
However the scene internal ids of resources change after moving. Is there any way to preserve them? (not critical for this PR though)

**Comment by Jordyfel** [comment]
Currently, I don't know what this depends on... I wish I could understand all the related systems so I can clean up these functions properly instead of continuing this chain of patch after patch, but that would take a while.

File: editor/filesystem_dock.cpp [hunk 3]
```diff
@@ -1553,50 +1558,30 @@ void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, Str
...
 
-void FileSystemDock::_update_dependencies_after_move(const HashMap<String, String> &p_renames) const {
+void FileSystemDock::_update_dependencies_after_move(const HashMap<String, String> &p_renames, const Vector<String> &p_remaps) const {
 	// The following code assumes that the following holds:
 	// 1) EditorFileSystem contains the old paths/folder structure from before the rename/move.
 	// 2) ResourceLoader can use the new paths without needing to call rescan.
-	Vector<String> remaps;
-	_find_remaps(EditorFileSystem::get_singleton()->get_filesystem(), p_renames, remaps);
-	for (int i = 0; i < remaps.size(); ++i) {
+	Vector<String> scenes_to_reload;
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:1567]:
```suggestion
	List<String> scenes_to_reload;
```
In cases where you only add at the end and don't need to access random elements, a list is better.

File: editor/filesystem_dock.cpp [hunk 4]
```diff
@@ -1553,50 +1558,30 @@ void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, Str
...
+				scenes_to_reload.push_back(file);
 			}
 		} else {
-			EditorNode::get_singleton()->add_io_error(TTR("Unable to update dependencies:") + "\n" + remaps[i] + "\n");
+			EditorNode::get_singleton()->add_io_error(TTR("Unable to update dependencies:") + "\n" + p_remaps[i] + "\n");
 		}
 	}
+
+	for (int i = 0; i < scenes_to_reload.size(); i++) {
+		EditorNode::get_singleton()->reload_scene(scenes_to_reload[i]);
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:1583]:
```suggestion
	for (const String &E : scenes_to_reload) {
		EditorNode::get_singleton()->reload_scene(E);
```

**Review by KoBeWi — APPROVED** [review]

**Comment by Jordyfel** [comment]
Also found #59116

**Review by YuriSizov — APPROVED** [review]
Looks reasonable enough to give it a try. Great job!

**Comment by YuriSizov** [comment]
Thanks! Let's see if anything falls off :)

**Pull request merged**